File: src/main/java/cs/ut/config/MasterConfiguration.java
package cs.ut.config;

import cs.ut.config.items.HeaderItem;
import cs.ut.config.items.ModelProperties;
import cs.ut.provider.ModelConfigurationProvider;
import cs.ut.provider.PageConfigurationProvider;
import org.apache.log4j.BasicConfigurator;
import org.apache.log4j.Logger;

import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBException;
import javax.xml.bind.Unmarshaller;
import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlElementWrapper;
import javax.xml.bind.annotation.XmlRootElement;
import java.io.File;
import java.util.List;

@XmlRootElement(name = "configuration")
public class MasterConfiguration {
    private static final Logger log = Logger.getLogger(MasterConfiguration.class);

    private static MasterConfiguration master;

    @XmlElement(name = "pageConfig")
    private PageConfigurationProvider pageConfigurationProvider;

    @XmlElement(name = "userLogDirectory")
    private String userLogDirectory;

    @XmlElementWrapper(name = "headerConfiguration")
    @XmlElement(name = "headerItem")
    private List<HeaderItem> headerItems;

    @XmlElement(name = "modelConfig")
    private ModelProperties modelProperties;

    private ModelConfigurationProvider modelConfigurationProvider;

    private MasterConfiguration() {
        configureLogger();
    }

    /**
     * Instantiates master configuration or if it has already been instantiated then returns the object
     *
     * @return master configuration object that contains configurable values defined in configuration.xml
     */
    public static MasterConfiguration getInstance() {
        if (master == null) {
            master = new MasterConfiguration();
            try {
                master.readMasterConfig();
            } catch (JAXBException e) {
                throw new RuntimeException("Failed to read master configuration", e);
            }
        }
        return master;
    }

    /**
     * Reads configuration.xml file and parses it into Java objects.
     *
     * @throws JAXBException if configuration is defined incorrectly
     */
    private void readMasterConfig() throws JAXBException {
        log.debug("Reading master configuration");
        File file = new File(getClass().getClassLoader().getResource("configuration.xml").getFile());

        JAXBContext jaxbContext = JAXBContext.newInstance(MasterConfiguration.class);
        Unmarshaller unmarshaller = jaxbContext.createUnmarshaller();
        MasterConfiguration configuration = (MasterConfiguration) unmarshaller.unmarshal(file);
        log.debug("Finished reading configuration");

        pageConfigurationProvider = configuration.getPageConfigurationProvider();
        log.debug(String.format("Successfully retrieved %s page configurations", pageConfigurationProvider.getPages().size()));

        userLogDirectory = configuration.getUserLogDirectory();
        log.debug(String.format("Successfully read user log directory: '%s'", userLogDirectory));

        headerItems = configuration.getHeaderItems();
        log.debug(String.format("Successfully read %s header items", headerItems.size()));

        modelProperties = configuration.getModelProperties();
        log.debug(String.format("Successfully read %s types and %s model parameters",
                modelProperties.getTypes().size(),
                modelProperties.getParameters().size()));

        modelConfigurationProvider = new ModelConfigurationProvider(modelProperties);

        log.debug("Successfully read master configuration");
    }

    public PageConfigurationProvider getPageConfigurationProvider() {
        return pageConfigurationProvider;
    }

    public String getUserLogDirectory() {
        return userLogDirectory;
    }

    public List<HeaderItem> getHeaderItems() {
        return headerItems;
    }

    private ModelProperties getModelProperties() {
        return modelProperties;
    }

    public ModelConfigurationProvider getModelConfigurationProvider() {
        return modelConfigurationProvider;
    }

    /**
     * Configures logger and Enables appenders for Log4j
     */
    private void configureLogger() {
        BasicConfigurator.configure();
    }
}

File: src/main/kotlin/cs/ut/config/MasterConfiguration.kt
package cs.ut.config

import cs.ut.config.items.ModelParameter
import cs.ut.config.nodes.*
import cs.ut.engine.JobManager
import cs.ut.jobs.StartUpJob
import cs.ut.logging.NirdLogger
import cs.ut.util.readHyperParameterJson
import org.apache.log4j.*
import java.io.File
import java.util.*
import javax.xml.bind.JAXBContext
import javax.xml.parsers.DocumentBuilder
import javax.xml.parsers.DocumentBuilderFactory
import kotlin.concurrent.timerTask

object MasterConfiguration {
    private val log = NirdLogger(caller = this.javaClass)

    val file: File = File(javaClass.classLoader.getResource("configuration.xml").file!!)

    lateinit var dirConfig: DirectoryConfiguration
    lateinit var modelConfiguration: ModelConfiguration

    lateinit var csvConfiguration: CsvConfiguration
    lateinit var threadPoolConfiguration: ThreadPoolConfiguration
    lateinit var headerConfiguration: HeaderConfiguration
    lateinit var pageConfiguration: PageConfiguration
    lateinit var userPreferences: UserPreferences
    lateinit var defaultValuesConfiguration: DefaultValuesConfiguration
    lateinit var tooltipConfig: TooltipConfig

    lateinit var optimizedParams: Map<String, List<ModelParameter>>

    init {
        readConfig()
        configureLogger()
        log.debug("Logger configured successfully")
        Timer().schedule(timerTask { JobManager.runServiceJob(StartUpJob()) }, 10000L)
    }

    /**
     * Configures logger and Enables appenders for Log4j
     */
    private fun configureLogger() {
        Logger.getRootLogger().removeAllAppenders()
        Logger.getRootLogger().additivity = false

        val ca = ConsoleAppender()
        ca.layout = PatternLayout("<%d{ISO8601}> <%p> <%C{1}.class:%L> <%m>%n")
        ca.threshold = Level.DEBUG
        ca.activateOptions()

        Logger.getRootLogger().addAppender(ca)

        val fileAppender = FileAppender()
        fileAppender.layout = PatternLayout("<%d{ISO8601}> <%p> <%C{1}.class:%L> <%m>%n")
        fileAppender.name = "nirdizati_ui_log.log"
        fileAppender.file = "nirdizati_ui_log.log"
        fileAppender.threshold = Level.DEBUG
        fileAppender.append = true
        fileAppender.activateOptions()

        Logger.getRootLogger().addAppender(fileAppender)
    }

    private fun <T> readClass(clazz: Class<T>, tag: String): T {
        log.debug("Reading class $clazz from configuration: $file from tag $tag")

        val dbf = DocumentBuilderFactory.newInstance()
        dbf.isNamespaceAware = true

        val db: DocumentBuilder = dbf.newDocumentBuilder()
        val doc = db.parse(file)
        log.debug("Successfully parsed document")

        val node = doc.getElementsByTagName(tag)
        log.debug("Successfully found node with tag $tag")

        val jaxbContext = JAXBContext.newInstance(clazz)
        return jaxbContext.createUnmarshaller().unmarshal(node.item(0), clazz).value
    }

    fun readConfig() {
        dirConfig = readClass(DirectoryConfiguration::class.java, "dirs")
        modelConfiguration= readClass(ModelConfiguration::class.java, "modelConfig")
            .also {
                it.prepareData()
            }

        csvConfiguration = readClass(CsvConfiguration::class.java, "csvConfig")
        threadPoolConfiguration = readClass(ThreadPoolConfiguration::class.java, "threadpool")
        headerConfiguration = readClass(HeaderConfiguration::class.java, "headerConfiguration")
        pageConfiguration = readClass(PageConfiguration::class.java, "pageConfig")
        userPreferences = readClass(UserPreferences::class.java, "userPreferences")
        defaultValuesConfiguration = readClass(DefaultValuesConfiguration::class.java, "defaultConfig")
        tooltipConfig = readClass(TooltipConfig::class.java, "tooltips")

        optimizedParams = readHyperParameterJson()
    }
}